### PR TITLE
#3008 Adding UI for participant's sourceIPs

### PR DESCRIPTION
**Pull request by SithuKyaw-Thitsaworks**

### CODE REVIEW

File: packages/admin-ui/package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
     "@mojaloop/security-bc-public-types-lib": "~0.2.1",
     "@mojaloop/settlements-bc-public-types-lib": "~0.2.2",
     "@ng-bootstrap/ng-bootstrap": "^10.0.0",
+    "@types/lodash": "^4.14.198",
```
Comment (pedrosousabarreto) [inline packages/admin-ui/package.json]: We should avoid gigantic libs like lodash/underscore, etc - removing it. 

This is **good enough** for deep copying objects without functions:
`const cloned = JSON.parse(JSON.stringify(object))`

Anyway, this import was only the types, the actual lib dependency was missing.

File: packages/admin-ui/src/app/participants/participant-detail.component.ts [hunk 2]
```diff
@@ -378,6 +372,178 @@ export class ParticipantDetailComponent implements OnInit {
...
+        break;
+      case "RANGE":
+        this.newSourceIp.ports = "";
+        break;
+      default:
+        break;
+    }
+  }
+
+  validateData(sourceIp: ParticipantAllowedSourceIp): boolean {
```
Comment (pedrosousabarreto) [inline packages/admin-ui/src/app/participants/participant-detail.component.ts]: This function needs a better name, like: isSourceIpValid()

File: packages/admin-ui/src/app/participants/participant-detail.component.ts [hunk 3]
```diff
@@ -378,6 +372,178 @@ export class ParticipantDetailComponent implements OnInit {
...
+      id: uuid.v4(),
+      allowedSourceIpId: sourceIp.id,
+      cidr: sourceIp.cidr,
+      portMode: sourceIp.portMode,
+      ports: sourceIp.portMode === "SPECIFIC" ? sourceIp.ports?.split(",").map(Number) : [],
+      portRange: sourceIp.portMode === "RANGE" ? {
+        rangeFirst: sourceIp.portRange.rangeFirst || 0,
+        rangeLast: sourceIp.portRange.rangeLast || 0
+      } :
+        { rangeFirst: 0, rangeLast: 0 },
```
Comment (pedrosousabarreto) [inline packages/admin-ui/src/app/participants/participant-detail.component.ts]: Never use magic values, like zeros or empty strings, when we can use undefined or null.

File: packages/admin-ui/src/app/participants/participant-detail.component.ts [hunk 4]
```diff
@@ -378,6 +372,178 @@ export class ParticipantDetailComponent implements OnInit {
     this.navBar.select("accounts");
   }
 
+  /*
+   * Source IPs
+   * */
+
+  onEditSourceIp(sourceIp: ParticipantAllowedSourceIp): void {
+    debugger
```
Comment (pedrosousabarreto) [inline packages/admin-ui/src/app/participants/participant-detail.component.ts]: We shouldn't leave debugger instructions when pushing the code

File: packages/admin-ui/src/app/participants/participant-detail.component.ts [hunk 5]
```diff
@@ -378,6 +372,178 @@ export class ParticipantDetailComponent implements OnInit {
...
+      } :
+        { rangeFirst: 0, rangeLast: 0 },
+      requestType: "ADD_SOURCE_IP"
+    }
+
+    console.log(sourceIpChangeRequest);
+
+    // check overlaps
+    if (this.sourceIpCreateModeEnabled) {
+      sourceIpChangeRequest.requestType = "ADD_SOURCE_IP";
```
Comment (pedrosousabarreto) [inline packages/admin-ui/src/app/participants/participant-detail.component.ts]: This duplicate check needs to happen also when editing

File: packages/admin-ui/src/app/participants/participant-detail.component.ts [hunk 6]
```diff
@@ -378,6 +372,178 @@ export class ParticipantDetailComponent implements OnInit {
...
+      const duplicateSourceIp = this.participant.value?.participantSourceIpChangeRequests?.find(
+        (item) =>
+          item.cidr.split('/')[0] === sourceIp.cidr.split('/')[0]
```
Comment (pedrosousabarreto) [inline packages/admin-ui/src/app/participants/participant-detail.component.ts]: We should compare the whole cidr

**Comment by SithuKyaw-Thitsaworks** [comment]
Thanks for the corrections and comments, Pedro.

**Pull request merged**